[Docs] upf documentation (#3723)
The-OpenROAD-Project/OpenROAD: src/upf/README.md

# Read UPF Utility

This module contains functionality to read, and modify information
from Unified Power Format (UPF) files. 

## Commands

```{note}
- Parameters in square brackets `[-param param]` are optional.
- Parameters without square brackets `-param2 param2` are required.
```

### Read UPF

Sources the UPF file. 

```tcl
read_upf

>     [-file file]

Reviewer: Shouldn't this be a required argument?

```

#### Options

| Switch Name | Description | 
| ----- | ----- |
| `-file` | `.upf` filename |

### Create Power Domain

```tcl
create_power_domain
    [-elements elements]
    name 
```

#### Options

| Switch Name | Description | 
| ----- | ----- |

> | `-elements` | list of module paths that belong this this domain OR `*` for top domain |
> | `name` | domain name |

Reviewer: ```suggestion
| `-elements` | List of module paths that belong to this domain. Alternatively, use the wildcard `*` for the top domain. |
| `name` | Domain name. |
```

### Create Logic Port

```tcl
create_logic_port
    [-direction direction]
    port_name
```

#### Options

| Switch Name | Description | 
| ----- | ----- |

> | `-direction` | direction of the port (`in`, `out`, `inout`) |
> | `port_name` | port name |

Reviewer: ```suggestion
| `-direction` | Direction of the port, one of (`in`, `out`, `inout`). |
| `port_name` | Port name. |
```

### Create Power Switch

```tcl
create_power_switch
    [-domain domain]
    [-output_supply_port output_supply_port]
    [-input_supply_port input_supply_port]
    [-control_port control_port]
    [-on_state on_state]
    name
```

#### Options

| Switch Name | Description | 
| ----- | ----- |
| `-domain` | Power domain name. |
| `-output_supply_port` | Output supply port of the switch. |
| `-input_supply_port` | Input supply port of the switch. |
| `-control_port` | Control port on the switch. |

> | `-on_state` | One of {`state_name` `input_supply_port` {`boolean_expression`}}. |

Reviewer: ```suggestion
| `-on_state` | One of {`state_name`, `input_supply_port`, `boolean_expression`}. |
```

> | `-domain` | power domain name |
> | `-output_supply_port` | output supply port of the switch |
> | `-input_supply_port` | input supply port of the switch |
> | `-control_port` | a control port on the switch |
> | `-on_state` | {`state_name` `input_supply_port` {`boolean_expression`}} |
> | `name` | power switch name |

Reviewer: ```suggestion
| `-domain` | Power domain name. |
| `-output_supply_port` | Output supply port of the switch. |
| `-input_supply_port` | Input supply port of the switch. |
| `-control_port` | Control port on the switch. |
| `-on_state` | One of {`state_name` `input_supply_port` {`boolean_expression`}}. |
| `name` | Power switch name. |
```

Reviewer: @luarss please capitalize the first word and add a `.` at the end of all sentences.

### Create or Update Isolation Strategy

```tcl 
set_isolation
    [-domain domain]
    [-applies_to applies_to]
    [-clamp_value clamp_value]
    [-isolation_signal isolation_signal]
    [-isolation_sense isolation_sense]
    [-location location]
    [-update]
    name
```

#### Options

| Switch Name | Description | 
| ----- | ----- |
| `-domain` | Power domain |
| `-applies_to` | Restricts the strategy to apply one of these (`inputs`, `outputs`, `both`). |

> | `clamp_value` | Value the isolation can drive (`0`, `1`). |
> | `isolation_signal` | The control signal for this strategy. |
> | `isolation_sense` | The active level of isolation control signal. |
> | `location` | Domain in which isolation cells are placed (`parent`, `self`, `fanout`). |
> | `update` | Flag if use existing strategy, will error if strategy doesn't exist. |

Reviewer: ```suggestion
| `-clamp_value` | Value the isolation can drive (`0`, `1`). |
| `-isolation_signal` | The control signal for this strategy. |
| `-isolation_sense` | The active level of isolation control signal. |
| `-location` | Domain in which isolation cells are placed (`parent`, `self`, `fanout`). |
| `-update` | Only available if using existing strategy, will error if the strategy doesn't exist. |
```

| `name` | Isolation strategy name. |

### Set Interface cell

```tcl 
use_interface_cell
    [-domain domain]
    [-strategy strategy]
    [-lib_cells lib_cells]
```

#### Options

| Switch Name | Description | 
| ----- | ----- |
| `-domain` | Power domain name. |
| `-strategy` | Isolation strategy name. |
| `-lib_cells` | List of lib cells that could be used. |

### Set Domain Area

```tcl
set_domain_area

>     {domain_name -area {llx lly urx ury}}

Reviewer: ```suggestion
    domain_name
    -area {llx lly urx ury}
```

```

#### Options

| Switch Name | Description | 
| ----- | ----- |

> | `domain_name` | Power domain name. |
> | `llx, lly, urx, ury` | The lower left and upper right x-/y- coordinates respectively of the power domain area (microns). |

Reviewer: ```suggestion
| `domain_name` | Power domain name. |
| `-area` | x-/y- coordinates in microns for the lower left and upper right corners of the power domain area. |
```
